akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+
+  val userHasData = new TransferState {
+    private val user = inputBunch.inputsOrCompleteAvailableFor(UserIn) || userInChoppingBlock
+    def isReady = !corkUser && user.isReady && lastHandshakeStatus != NEED_UNWRAP
+    def isCompleted = false
+  }
+
+  val transportHasData = inputBunch.inputsOrCompleteAvailableFor(TransportIn) || transportInChoppingBlock
+
+  // bidirectional case
+  val outbound = (userHasData || engineNeedsWrap) && outputBunch.demandAvailableFor(TransportOut)
+  val inbound = transportHasData && outputBunch.demandOrCancelAvailableFor(UserOut)
+
+  // half-closed
+  val outboundHalfClosed = engineNeedsWrap && outputBunch.demandAvailableFor(TransportOut)
+  val inboundHalfClosed = transportHasData && engineInboundOpen
+
+  def completeOrFlush(): Unit =
+    if (engine.isOutboundDone()) nextPhase(completedPhase)
+    else nextPhase(flushingOutbound)
+
+  private def doInbound(isOutboundClosed: Boolean, inboundState: TransferState): Boolean =
+    if (inputBunch.isDepleted(TransportIn) && transportInChoppingBlock.isEmpty) {
+      if (tracing) log.debug("closing inbound")
+      try engine.closeInbound()
+      catch { case ex: SSLException ⇒ outputBunch.enqueue(UserOut, SessionTruncated) }
+      completeOrFlush()
+      false
+    } else if (inboundState != inboundHalfClosed && outputBunch.isCancelled(UserOut)) {
+      if (!isOutboundClosed && closing.ignoreCancel) {
+        if (tracing) log.debug("ignoring UserIn cancellation")
+        nextPhase(inboundClosed)
+      } else {
+        if (tracing) log.debug("closing inbound due to UserOut cancellation")
+        engine.closeOutbound()
+        nextPhase(flushingOutbound)
+      }
+      true
+    } else if (inboundState.isReady) {
+      transportInChoppingBlock.chopInto(transportInBuffer)
+      try {
+        doUnwrap()
+        true
+      } catch {
+        case ex: SSLException ⇒
+          if (tracing) log.debug(s"SSLException during doUnwrap: $ex")
+          completeOrFlush()
+          false
+      }
+    } else true
+
+  private def doOutbound(isInboundClosed: Boolean): Unit =
+    if (inputBunch.isDepleted(UserIn) && userInChoppingBlock.isEmpty) {
+      if (!isInboundClosed && closing.ignoreComplete) {
+        if (tracing) log.debug("ignoring closeOutbound")
+      } else {
+        if (tracing) log.debug("closing outbound directly")
+        engine.closeOutbound()
+      }
+      nextPhase(outboundClosed)
+    } else if (outputBunch.isCancelled(TransportOut)) {
+      nextPhase(completedPhase)
+    } else if (outbound.isReady) {
+      if (userHasData.isReady) userInChoppingBlock.chopInto(userInBuffer)
+      try doWrap()
+      catch {
+        case ex: SSLException ⇒
+          if (tracing) log.debug(s"SSLException during doWrap: $ex")
+          completeOrFlush()
+      }
+    }
+
+  val bidirectional = TransferPhase(outbound || inbound) { () ⇒
+    if (tracing) log.debug("bidirectional")
+    val continue = doInbound(isOutboundClosed = false, inbound)
```

Comment:
A solution (in user space) would be a bunching stage in front of UserIn that just groups things into >16kiB chunks.